tests/cycle.rs

```diff
@@ -0,0 +1,881 @@
+    let a = Input::MinIterate(a_in);
+    let b = Input::MinPanic(b_in);
+    a_in.set_inputs(&mut db).to(vec![b]);
+    b_in.set_inputs(&mut db).to(vec![value(30)]);
+
+    a.clone().assert_value(&db, 30);
+
+    b_in.set_inputs(&mut db).to(vec![a.clone()]);
+
+    a.assert_value(&db, 255);
+}
+
+/// a:Xi(b) -> b:Xi(c, a) -> c:Xp(v25, a, sb)
+/// ^          ^        |                   |
+/// +----------+--------|-------------------+
+/// |                   |
+/// +-------------------+
+///
+/// Nested cycles, double head. We hit max iterations and fall back, then max value on the next
+/// revision, then converge on the next.
+#[test]
+fn nested_double_multiple_revisions() {
+    let mut db = DbImpl::new();
+    let a_in = Inputs::new(&db, vec![]);
+    let b_in = Inputs::new(&db, vec![]);
+    let c_in = Inputs::new(&db, vec![]);
+    let a = Input::MaxIterate(a_in);
+    let b = Input::MaxIterate(b_in);
+    let c = Input::MaxPanic(c_in);
+    a_in.set_inputs(&mut db).to(vec![b.clone()]);
+    b_in.set_inputs(&mut db).to(vec![c, a.clone()]);
+    c_in.set_inputs(&mut db).to(vec![
+        value(25),
+        a.clone(),
+        Input::Successor(Box::new(b.clone())),
+    ]);
+
+    a.clone().assert_count(&db);
+
+    // next revision, we hit max value instead
+    c_in.set_inputs(&mut db).to(vec![
+        value(243),
+        a.clone(),
+        Input::Successor(Box::new(b.clone())),
+    ]);
+
+    a.clone().assert_bounds(&db);
+
+    // and next revision, we converge
+    c_in.set_inputs(&mut db).to(vec![value(240), a.clone(), b]);
+
+    a.assert_value(&db, 240);
+}
+
+/// a:Ni(b) -> b:Ni(c) -> c:Ni(a)
+/// ^                           |
+/// +---------------------------+
+///
+/// In a cycle with some LOW durability and some HIGH durability inputs, changing a LOW durability
+/// input still re-executes the full cycle in the next revision.
+#[test]
+fn cycle_durability() {
+    let mut db = DbImpl::new();
+    let a_in = Inputs::new(&db, vec![]);
+    let b_in = Inputs::new(&db, vec![]);
+    let c_in = Inputs::new(&db, vec![]);
+    let a = Input::MinIterate(a_in);
+    let b = Input::MinIterate(b_in);
+    let c = Input::MinIterate(c_in);
+    a_in.set_inputs(&mut db)
+        .with_durability(Durability::LOW)
+        .to(vec![b.clone()]);
+    b_in.set_inputs(&mut db)
+        .with_durability(Durability::HIGH)
+        .to(vec![c]);
+    c_in.set_inputs(&mut db)
+        .with_durability(Durability::HIGH)
+        .to(vec![a.clone()]);
+
+    a.clone().assert_value(&db, 255);
+
+    // next revision, we converge instead
+    a_in.set_inputs(&mut db)
+        .with_durability(Durability::LOW)
+        .to(vec![value(45), b]);
+
+    a.assert_value(&db, 45);
+}
+
+/// a:Np(v59, b) -> b:Ni(v60, c) -> c:Np(b)
+///                 ^                     |
+///                 +---------------------+
+///
+/// If nothing in a cycle changed in the new revision, no part of the cycle should re-execute.
+#[test]
+fn cycle_unchanged() {
```

Comment:
Not required but it could be nice to have a multi threaded tests where:

* the recovery function panics
* The cycle head panics while iterating

Mainly to show that we don't dead-lock the other thread.